docs/weave/design/03-permissions.md

```diff
@@ -0,0 +1,156 @@
+
+### Serialization
+
+Now we have a clear understanding of what conditions are in this context, and that there may be different conditions on one object, we need to consider how to store them in the database. A condition can be considered a tuple of `(extension, type, data)`, for example a ed25519 public key signature could be represented as `("sigs", "ed25519", <addr>)` and a sha256 hashlock could be `("hash", "sha256", <hash>)`.
+
+Note that the "data" doesn't need to reveal what the data is that will match this condition, but needs to be calculated from it (eg. addr is first 20 bytes of a hash of the public key). And each extension and type may have different interpretations of the data.
+
+If we enforce simple text for extension and type, we could encode it as `sprintf("%s/%s/%X", extension, type, data)`. This is longer than the 20 bytes often used for addresses, and maybe we could hash it first, but then we loose information. I can envision a user wanting to know if an account is controlled by a private key, a hash preimage, or another contract. If I am going to set up an escrow with the same arbiter as you on another chain to do atomic swap, I want to make sure that it is controlled by a hash preimage (which you must reveal), not your private key (which would not let me collect the other escrow).
+
+### Addresses
+
+We started with a simple address function, which was the first 20 bytes of the sha256 hash of a public key. However, this left no room for other authentication mechanisms, and it wasn't even clear how we could differentiate between ed25519 and secp256k1 signatures. However, the usecase was clear: a short identifier that was uniquely tied to an authentication condition, but *did not reveal* that condition.
+
+We can redefine address to be the hash of a "condition", not just public key bytes, and then we keep this functionality while generalizing what a condition is.
+
+```go
+condition := sprintf("%s/%s/%X", extension, type, data)
+address := sha256(condition)[:20]
+```
+
+The questions is when and how to use each one. Any field that can declare an owner must decide if those bytes represent a condition or an address. The Authenticator can store fulfilled Conditions in the Context, and then allow clients to check for matches either by condition or by address. But where to use which one?
+
+Here are some rough guidelines:
+
+1. If we really need to save 20 bytes, use an *Address*. (But few places need that micro-optimization)
+2. If we need visibility of control, use *Condition* (multi-sig solutions, arbiters, etc)
+3. If you want to obscure control (until first use), use *Address*
+4. Everything else, at your discretion, but prefer *Address* when possible for consistency.
+
+I guess it is up to the extension developer, but I would generally use Conditions for anything stored in the value and Address for fields that appear in the key, unless there is a reason otherwise.
```

Comment:
```suggestion
While it is a descision for the extension developer, we would generally use Conditions for anything stored in the value and Address for fields that appear in the key, unless there is a reason to do otherwise.
```